utils/orderedset.go

```diff
@@ -0,0 +1,69 @@
+package utils
+
+import (
+	"sync"
+)
+
+// OrderedSet is a thread-safe data structure that maintains both uniqueness and insertion order of elements.
+// It combines the benefits of both maps and slices:
+// - Uses a map for O(1) lookups and to ensure element uniqueness
+// - Uses a slice to maintain insertion order and enable ordered iteration
+// The data structure is safe for concurrent access through the use of a read-write mutex.
+type OrderedSet[K comparable, V any] struct {
+	nodeSet  map[K]V
```

Comment:
It looks like we can go with 
```go
nodeSet map[K]struct{}
```
to save memory and we don't keep value twice. WDYT?

Reply:
Ah actually we can remove `nodeSet` completely, because we already have `nodePos` to track the position of the item in `nodeList`.